events/diskpersist.go

```diff
@@ -1,19 +1,447 @@
+	}
+
+	flags := binary.LittleEndian.Uint32(scratch[:4])
+	kind := binary.LittleEndian.Uint32(scratch[4:8])
+	l := binary.LittleEndian.Uint32(scratch[8:12])
+	seq := binary.LittleEndian.Uint64(scratch[12:20])
+
+	return &evtHeader{
+		Flags: flags,
+		Kind:  kind,
+		Len:   int64(l),
+		Seq:   int64(seq),
 	}, nil
 }
 
-func (p *DiskPersistence) Persist(e *XRPCStreamEvent) error {
-	panic("nyi")
+func (p *DiskPersistence) writeHeader(ctx context.Context, flags uint32, kind uint32, l uint32, seq int64) error {
+	binary.LittleEndian.PutUint32(p.scratch, flags)
+	binary.LittleEndian.PutUint32(p.scratch[4:], kind)
+	binary.LittleEndian.PutUint32(p.scratch[8:], l)
+	binary.LittleEndian.PutUint64(p.scratch[12:], uint64(seq))
+
+	nw, err := p.logfi.Write(p.scratch)
+	if err != nil {
+		return err
+	}
+
+	if nw != headerSize {
+		return fmt.Errorf("only wrote %d bytes for header", nw)
+	}
+
+	return nil
+}
+
+func (p *DiskPersistence) uidForDid(ctx context.Context, did string) (util.Uid, error) {
+	if uid, ok := p.didCache.Get(did); ok {
+		return uid.(util.Uid), nil
+	}
+
+	var u models.ActorInfo
+	if err := p.meta.First(&u, "did = ?", did).Error; err != nil {
+		return 0, err
+	}
+
+	p.didCache.Add(did, u.Uid)
+
+	return u.Uid, nil
+}
+
+func (p *DiskPersistence) Playback(ctx context.Context, since int64, cb func(*XRPCStreamEvent) error) error {
+	var logs []LogFileRef
+	if err := p.meta.Debug().Order("seq_start asc").Find(&logs, "seq_start >= ?", since%p.eventsPerFile).Error; err != nil {
+		return err
+	}
+
+	for _, lf := range logs {
+		fmt.Println("LOGS", lf)
+		if err := p.readEventsFrom(ctx, since, filepath.Join(p.primaryDir, lf.Path), cb); err != nil {
+			fmt.Println("EVENT ERROR: ", err)
+			return err
+		}
+		since = 0
+	}
+
+	return nil
+}
+
+func (p *DiskPersistence) readEventsFrom(ctx context.Context, since int64, fn string, cb func(*XRPCStreamEvent) error) error {
+	fi, err := os.OpenFile(fn, os.O_RDONLY, 0)
+	if err != nil {
+		return err
+	}
+
+	if since != 0 {
+		_, err := scanForLastSeq(fi, since)
+		if err != nil {
+			return err
+		}
+	}
+
+	bufr := bufio.NewReader(fi)
+
+	scratch := make([]byte, headerSize)
+	for {
+		h, err := readHeader(bufr, scratch)
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				return nil
+			}
+
+			return err
+		}
+
+		switch h.Kind {
+		case evtKindCommit:
+			var evt atproto.SyncSubscribeRepos_Commit
+			if err := evt.UnmarshalCBOR(io.LimitReader(bufr, h.Len)); err != nil {
+				return err
+			}
+			evt.Seq = h.Seq
+			if err := cb(&XRPCStreamEvent{RepoCommit: &evt}); err != nil {
+				return err
+			}
+		case evtKindHandle:
+			var evt atproto.SyncSubscribeRepos_Handle
+			if err := evt.UnmarshalCBOR(io.LimitReader(bufr, h.Len)); err != nil {
+				return err
+			}
+			evt.Seq = h.Seq
+			if err := cb(&XRPCStreamEvent{RepoHandle: &evt}); err != nil {
+				return err
+			}
+		default:
+			log.Warnw("unrecognized event kind coming from log file", "seq", h.Seq, "kind", h.Kind)
```

Comment:
Would be nice to put the byte offset for the log file in here so we can investigate if that's easy